### PR TITLE
Pet Set Example for Cassandra

### PR DESCRIPTION
**Pull request by chrislovecnm**
- updating cassandra to 3.7
- added pet set example
- adding pet set for Cassandra e2e tests
- changed service as we do not want a lb service, as we are running C*
- updated docs

cc @bgrant0607 

cc @kubernetes/examples 

We can probably close a couple of other open PR, since I did some other stuff.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30577)

### CODE REVIEW

**Comment by chrislovecnm** [comment]
@bgrant0607 

BTW ... I NEED THE DOCKER UPDATED ... do not test this will break the e2e tests....  Please loop in   someone who can update the docker image.

@bprashanth / @smarterclayton  can you look at the e2e test?

Should resolve

https://github.com/kubernetes/kubernetes/issues/27476
https://github.com/kubernetes/kubernetes/issues/25889
https://github.com/kubernetes/kubernetes/issues/24030
https://github.com/kubernetes/kubernetes/issues/28835
https://github.com/kubernetes/kubernetes/issues/27075

Initial support for

https://github.com/kubernetes/kubernetes/issues/24286

I need to proof the docs more, but have at it wordsmiths!

File: examples/storage/cassandra/cassandra-petset.yaml
Comment (ram-argus) [inline examples/storage/cassandra/cassandra-petset.yaml]: @chrislovecnm what is  the reasom for the "priviledged" security setting?
any chance to consider something more specific:
```
          securityContext:
            capabilities:
              add:
                - NET_ADMIN
```

Comment (chrislovecnm) [inline examples/storage/cassandra/cassandra-petset.yaml]: In order for the jvm to reserve memory with the kernel you need this.

Comment (chrislovecnm) [inline examples/storage/cassandra/cassandra-petset.yaml]: @ram-argus is there an option that will allow the jvm to do a mockall without running as privledged? I would have to mess with it, but this should not be a change.

**Comment by chrislovecnm** [comment]
I have zero idea what happened to my documentation.  git ate my homework. ARG  ... will get it checked in when and if I find it.

**Comment by rata** [comment]
@chrislovecnm if you commited and lost it in some rebase, you can probably find it with `git reflog`

**Comment by chrislovecnm** [comment]
@rata figured that out last night ... gotta did it out of history

File: examples/storage/cassandra/cassandra-petset.yaml
Comment (ram-argus) [inline examples/storage/cassandra/cassandra-petset.yaml]: @chrislovecnm looks like IPC_LOCK does that [(runtime-privilege-and-linux-capabilities)](https://docs.docker.com/engine/reference/run/#/runtime-privilege-and-linux-capabilities)
```
          securityContext:
            capabilities:
              add:
                - IPC_LOCK
```

File: docs/user-guide/petsets.md
Comment (jeffmendoza) [inline docs/user-guide/petsets.md]: Do you need to add this empty doc?

Comment (chrislovecnm) [inline docs/user-guide/petsets.md]: This is WIP because git ate my documentation.  The problem is probably between the floor and the keyboard, but for now, I will blame it on git.  I am going to dig it out with reflog later today.

**Comment by chrislovecnm** [comment]
So the documentation has been revived.   Please let me know what you think.

**Comment by chrislovecnm** [comment]
@ram-argus I still need to test the security context that you recommended

**Comment by jeffmendoza** [comment]
@chrislovecnm Looks like the docs are still not quite right.

**Comment by chrislovecnm** [comment]
@jeffmendoza @smarterclayton @bprashanth a coupe of things

update-munge-docs.sh is completely destroying my README:

```console
$ hack/update-munge-docs.sh
+++ [0822 11:37:16] Generating bindata:
    /Users/clove/Workspace/src/github.com/kubernetes/kubernetes/test/e2e/framework/gobindata_util.go
+++ [0822 11:37:17] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [0822 11:37:17] Building go targets for darwin/amd64:
    cmd/mungedocs
/Users/clove/Workspace/src/github.com/kubernetes/kubernetes/examples/storage/cassandra/README.md
----
md-links:
contents were modified

blank-lines-surround-preformatted:
contents were modified

sync-examples:
never found closing end mark while updating macro blocks

FAIL: some manual changes are still required.
/Users/clove/Workspace/src/github.com/kubernetes/kubernetes/examples/ requires manual changes.  See preceding errors.
```

verify is not giving me much feedback.

Can someone build and deploy the new docker?  I need to test that and the memory / security tweak that was recommended.

File: docs/user-guide/petsets.md
Comment (jeffmendoza) [inline docs/user-guide/petsets.md]: These were here due to the docs being moved to k8s.io, and not wanting to break existing links. I don't think we need new empty redirect docs.

File: examples/storage/cassandra/image/Makefile
Comment (jeffmendoza) [inline examples/storage/cassandra/image/Makefile]: Why is the jar not built as part of the Docker build?

**Comment by jeffmendoza** [comment]
@chrislovecnm `gcr.io/google-samples/cassandra:v11` pushed. I did not build the jar, just used the one you have checked in.

File: examples/storage/cassandra/image/Makefile
Comment (chrislovecnm) [inline examples/storage/cassandra/image/Makefile]: Wonder if we should even have the jar checked in.

**Comment by chrislovecnm** [comment]
@jeffmendoza @bprashanth any idea why I am getting an error in the e2e tests that I don't have a petset e2e test, when I have one here: https://github.com/chrislovecnm/kubernetes/blob/a666ea9da07c1c3bb0eb001ac32cd927c7442115/test/e2e/examples.go#L306

**Comment by jeffmendoza** [comment]
@chrislovecnm examples_test.go is not related to e2e, and needs it's own test. Add a line for cassandra-petset.yaml here: https://github.com/chrislovecnm/kubernetes/blob/a666ea9da07c1c3bb0eb001ac32cd927c7442115/examples/examples_test.go#L259-L263

**Comment by chrislovecnm** [comment]
@jeffmendoza updated for tests and the sercurityContext.  The new docker images is working like a champ.  Anything else?

**Comment by chrislovecnm** [comment]
@jeffmendoza looks like a flake or another unit test boomed.  Recommendations?

**Comment by jeffmendoza** [comment]
@chrislovecnm You can try to rebase (and squash) and push --force, which will re-trigger tests.

**Comment by chrislovecnm** [comment]
@jeffmendoza oh joy.  Sorry last time I did a squash there were problems between the floor and the keyboard.

**Comment by jeffmendoza** [comment]
First do a merge, it is easier to resolve conflicts that way (although Github says it merges cleanly).
Then do `git rebase -i origin/master`, change the first line to `reword` and all the others to `squash`

**Comment by chrislovecnm** [comment]
@jeffmendoza I was actually able to rebase and a squash.  Fingers crossed.

**Comment by jeffmendoza** [comment]
LGTM

**Comment by chrislovecnm** [comment]
@jeffmendoza makes me happy ;). Thanks! Now I gotta close a bunch of issues.

**Comment by chrislovecnm** [comment]
@jeffmendoza you did a LGTM, and Submit Queue did not pick it up ... DOH

**Comment by foxish** [comment]
The comment needs to be `/lgtm` in case you want the submit-queue to auto-apply the label and pick it up. https://github.com/kubernetes/contrib/blob/master/mungegithub/README.md

**Comment by jeffmendoza** [comment]
@foxish Oh, I don't have permissions.

**Comment by foxish** [comment]
@jeffmendoza I don't think it would need any special permission. The reason we have the `/lgtm` comment is to allow folks who cannot apply labels to still review things. So, at this point, if you were to comment with a lone `/lgtm`, that would work. I'll apply the label for now.

**Comment by foxish** [comment]
Not sure about the release notes here. @bprashanth, does this need a release note?

**Comment by jeffmendoza** [comment]
Ah cool! I didn't know.

**Comment by chrislovecnm** [comment]
@foxish yes this should have RL ... I am bumping versions ... blah blah blah

**Comment by chrislovecnm** [comment]
Review status: 0 of 22 files reviewed at latest revision, 3 unresolved discussions.

---

*[examples/storage/cassandra/image/Makefile, line 25 \[r6\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30577#-KP51czkaf7Je3N92fMj-r6-25:-KRKXpwjgarAAS3iTyGY:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/a666ea9da07c1c3bb0eb001ac32cd927c7442115/examples/storage/cassandra/image/Makefile#L25)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary>
> Wonder if we should even have the jar checked in.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30577)*

**Comment by chrislovecnm** [comment]
Review status: 0 of 22 files reviewed at latest revision, 3 unresolved discussions.

---

*[docs/user-guide/petsets.md, line 34 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30577#-KP51czkaf7Je3N92fMr-r2-34:-KRKYomkK_i8kVI9oc9c:b-ul84tf) ([raw file](https://github.com/kubernetes/kubernetes/blob/552b5118de41fdc5de252fa72ae09fbb3c65c295/docs/user-guide/petsets.md#L34)):*
<details><summary><i>Previously, jeffmendoza (Jeff Mendoza) wrote…</i></summary>
> These were here due to the docs being moved to k8s.io, and not wanting to break existing links. I don't think we need new empty redirect docs.
</details>
I think these are picked up by the doc checking tools we have.  This is actually holding up this PR.  Comments on Renewable​.

---

*[examples/storage/cassandra/image/Makefile, line 25 \[r6\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30577#-KP51czkaf7Je3N92fMj-r6-25:-KRKYXTWsKvf1fHLPVT8:b-tbc9wh) ([raw file](https://github.com/kubernetes/kubernetes/blob/a666ea9da07c1c3bb0eb001ac32cd927c7442115/examples/storage/cassandra/image/Makefile#L25)):*
<details><summary><i>Previously, chrislovecnm (Chris Love) wrote…</i></summary>
> Done.
</details>
We have a bigger problem with docker dependencies and these examples.  There is a bigger topic from this

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30577)*

**Comment by chrislovecnm** [comment]
Review status: 0 of 22 files reviewed at latest revision, 3 unresolved discussions.

---

*[examples/storage/cassandra/cassandra-petset.yaml, line 36 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30577#-KP51czkaf7Je3N92fMs-r2-36:-KRKZ1QrttVjz_K0V4Kn:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/552b5118de41fdc5de252fa72ae09fbb3c65c295/examples/storage/cassandra/cassandra-petset.yaml#L36)):*
<details><summary><i>Previously, ram-argus wrote…</i></summary>
> @chrislovecnm looks like IPC_LOCK does that [(runtime-privilege-and-linux-capabilities)](https://docs.docker.com/engine/reference/run/#/runtime-privilege-and-linux-capabilities)
> 
> ```
> 
>           securityContext:
> 
>             capabilities:
> 
>               add:
> 
>                 - IPC_LOCK
> 
> ```
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30577)*

**Comment by chrislovecnm** [comment]
@jeffmendoza and @ram-argus comments on reviewable are holding up this PR.  Thoughts?

https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30577

**Comment by foxish** [comment]
@chrislovecnm It's not the comments. It is the do-not-merge label. Removing it now, since @jeffmendoza already LGTM-ed it.

**Comment by chrislovecnm** [comment]
@foxish much thanks, let me know, as I would like to get this put to bed

**Comment by foxish** [comment]
@chrislovecnm I guess the submit-queue doesn't want to auto-merge this. I'll get this sorted out on Monday.

**Comment by chrislovecnm** [comment]
@foxish :rimshot ... thanks

**Comment by rmmh** [comment]
@k8s-bot gke test this issue #32431

**Comment by chrislovecnm** [comment]
@rmmh thanks for the follow-up

**Comment by foxish** [comment]
@eparis Looks like the block-paths munger doesn't allow this to get auto-merged. 
What is the right procedure around this?

**Comment by chrislovecnm** [comment]
@foxish any ideas?

**Comment by chrislovecnm** [comment]
sorry to be a bit pushy guys, but this has been open awhile.  Does this need to wait till after the release?

cc: @bprashanth @smarterclayton @bgrant0607

**Comment by foxish** [comment]
@chrislovecnm Looks like this won't be automerged till you remove `docs/user-guide/petsets.md`. It falls under old-docs which shouldn't be changed anymore, since we've moved to the separate kubernetes.github.io repository, and hence, is blocked by the submit-queue.

**Comment by chrislovecnm** [comment]
So how does the link checker work then, maybe it has changed. Will pull the doc from the PR.

**Comment by chrislovecnm** [comment]
@foxish much thanks.  Removed.  Now lets see if we can get this in!

File: examples/storage/cassandra/README.md
Comment (foxish) [inline examples/storage/cassandra/README.md]: The path is invalid now, needs to be changed to the main documentation.

File: examples/storage/cassandra/README.md
Comment (foxish) [inline examples/storage/cassandra/README.md]: Same here, invalid path. Can point to http://kubernetes.io/docs/user-guide/petset/ ?

File: examples/storage/cassandra/README.md
Comment (foxish) [inline examples/storage/cassandra/README.md]: Same comment as above.

**Review by foxish — CHANGES_REQUESTED** [review]

**Comment by foxish** [comment]
@chrislovecnm mungedocs finds those links broken and they'll need to be fixed.

**Comment by chrislovecnm** [comment]
@foxish ran the wrong verify tool, yay ...  Will fix.

File: examples/storage/cassandra/README.md
Comment (chrislovecnm) [inline examples/storage/cassandra/README.md]: Done.

File: examples/storage/cassandra/README.md
Comment (chrislovecnm) [inline examples/storage/cassandra/README.md]: Done

File: examples/storage/cassandra/README.md
Comment (chrislovecnm) [inline examples/storage/cassandra/README.md]: And done.

**Review by foxish — APPROVED** [review]

**Comment by foxish** [comment]
@k8s-bot test this please, issue #IGNORE

**Comment by foxish** [comment]
@k8s-bot test this please, issue #IGNORE

**Comment by VinceMD** [comment]
Trying to figure out where things are at. Sorry this ticket is closed but not sure many people are looking at the Examples repository. 
I have some time in hands to test and document on an AWS setup across 3 zones (racks/AZ). Not looking at cross regions/DC at the moment.

I'm happy to create new tickets, just tell me where. Examples or here?

1. *EC2 snitch*: has anyone tested it in the end? If not, I can test it. I just need some guidance on what to do, how to get started
2. *Kubernetes snitch*: I understand the work has not started yet
3. *Kubernetes custom seed provider*: It looks like the code is there but is not enabled in the example and not included in the docker image. See my ticket https://github.com/kubernetes/examples/issues/147
4. *Kubernetes + 1 Cloud zone down + statefulset + PVC constraint = no pod can be fixed/recreated*: https://github.com/kubernetes/examples/issues/149

**Pull request merged**
